uimaj-core/src/main/java/org/apache/uima/cas/impl/BinaryCasSerDes.java

```diff
@@ -1697,12 +1700,23 @@ private void createFSsFromHeaps(boolean isDelta, int startPos, CommonSerDesSeque
                // start of normal non-array
         CASImpl view = null; 
         boolean isSofa = false;
+        boolean documentAnnotationPreviouslyIndexed = false;
         if (type.isAnnotationBaseType()) {
           Sofa sofa = getSofaFromAnnotBase(heapIndex, stringHeap, addr2fs, csds);  // creates sofa if needed and exists (forward ref case)
           view = (sofa == null) ? baseCas.getInitialView() : baseCas.getView(sofa);
           if (type == tsi.docType) {
-            fs = view.getDocumentAnnotation();  // creates the document annotation if it doesn't exist
-            view.removeFromCorruptableIndexAnyView(fs, view.getAddbackSingle());
+            Annotation documentAnnotationPrevious = view.getDocumentAnnotationNoCreate();
+            if (documentAnnotationPrevious == null) {
+              // document annotation not present
+              fs = view.createDocumentAnnotationNoRemoveNoIndex(0); // create but don't index
+              view.set_deserialized_doc_annot_not_indexed((Annotation)fs);  // for use by other code that sets length, if this is not indexed
+              // documentAnnotationPreviouslyIndex == false, preset above
+            } else {
+                fs = documentAnnotationPrevious;  

```

Comment:
I believe there is no code which prevents the user from manually instantiating multiple document annotation instances and adding these to the indexes. Would the code support that or mess up?

Reply:
I think the code mostly ignores this.  There's a method in the cas to get the DocumentAnnotation, and it runs an annotation iterator over that type in the particular view, and uses the 1st one.   I'm pretty sure no one has thought about that case, though.